### PR TITLE
Port backend tests to testify / fix racy tests

### PR DESCRIPTION
**Pull request by tcsc**
In order to get better visibility into the backend database tests using
the standard Go tooling, this changeset ports the backend tests away
from `Check`, and into subtests & `testify` for assertions.

This change means that individual sub-tests
 1. can be more easily identified in the json test logs, and
 2. can be more easily run individually from the command line

It also cuts down the amount of boilerplate required for each backend test.

During this port I also discovered that some tests are using the fake
clocks incorrectly, which may be a cause of some of our flaky `etcd`
tests.

I've also added as much commentary as I can deduce from context to try and make it easier to reason about the test intent.

### CODE REVIEW

File: lib/backend/test/suite.go [hunk 1]
```diff
@@ -547,235 +651,310 @@ func (s *BackendSuite) Locking(c *check.C, bk backend.Backend) {
...
+	item, err = uut.Get(ctx, prefix("a"))
+	require.NoError(t, err)
+	require.Equal(t, originalID, item.ID)
 
 	// Add item to backend that is already expired.
 	item2 := &backend.Item{
 		Key:     prefix("b"),
 		Value:   []byte("val"),
-		Expires: time.Now().Add(-1 * time.Second),
+		Expires: uut.Clock().Now().Add(-1 * time.Second),
```
Comment (tcsc) [inline lib/backend/test/suite.go:920]: Note clock change

File: lib/backend/test/suite.go [hunk 2]
```diff
@@ -547,235 +651,310 @@ func (s *BackendSuite) Locking(c *check.C, bk backend.Backend) {
...
-		c.Fatalf("Timeout waiting for event.")
-	}
+	requireEvent(t, watcher, types.OpInit, nil, 2*time.Second)
 
 	// Add item to backend with a 1 second TTL.
 	item := &backend.Item{
 		Key:     prefix("a"),
 		Value:   []byte("val"),
-		Expires: time.Now().Add(1 * time.Second),
+		Expires: uut.Clock().Now().Add(1 * time.Second),
```
Comment (tcsc) [inline lib/backend/test/suite.go:875]: Note clock change. Old code was taking timestamp from system clock then comparing against fake testing clock.

**Comment by tcsc** [comment]
Reviewers, I recommend starting at `lib/backend/test/suite.go` and working back from there.

**Review by russjones — APPROVED** [review]
Bot. Ready to merge whenever reviewers feel comfortable.

File: lib/backend/test/suite.go [hunk 3]
```diff
@@ -21,90 +21,212 @@ package test
...
+	result := ConstructionOptions{
+		MirrorMode: false,
+	}
+	if err := result.Apply(options); err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
+
+// applyOptions applies a collection of option-setting functions to the
```
Comment (quinqu) [inline lib/backend/test/suite.go:67]: Nit:
```suggestion
// Apply applies a collection of option-setting functions to the
```

Comment (quinqu) [reply lib/backend/test/suite.go:67]: or maybe you want to actually rename this back to `applyOptions` so it mimics the public function it's being called in

**Review by quinqu — APPROVED** [review]

File: lib/backend/test/suite.go (hunk 2 above)
Comment (quinqu) [reply lib/backend/test/suite.go:892]: Nice catch

**Review by xacrimon — APPROVED** [review]

**Pull request merged**